src/measurement_kit/cmdline/agent.cpp

```
@@ -0,0 +1,305 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include <measurement_kit/ext.hpp>
+#include <measurement_kit/ooni.hpp>
+
+#include "../cmdline/cmdline.hpp"
+/*
+ * TODO: made these interfaces available through measurement-kit public API
+ */
+#include "../../libmeasurement_kit/libevent/connection.hpp"
+#include "../../libmeasurement_kit/libevent/listen.hpp"
+
+namespace mk {
+namespace cmdline {
+namespace agent {
+
+/*
+ * FIXME With this usage, anyone can see the password using `ps auxww`
+ */
+#define USAGE "usage: measurement_kit agent [-v] [-p port] passwd\n"
+
+class Context {
+  public:
+    bool authenticated = false;
+    Var<net::Buffer> buff{new net::Buffer};
+    Var<Logger> logger;
+    std::string passwd = "";
+    Var<Reactor> reactor;
+    Var<net::Transport> txp;
+};
+
+/**
+ * Remove the next token from string.
+ * @param s String to read token from, updated in output.
+ * @return Next token or error.
+ */
+static ErrorOr<std::string> remove_next_token(std::string &s) {
+    size_t pos = s.find_first_of(" ");
+    if (pos == std::string::npos) {
+        return GenericError();
```

TODO: here we should probably return a better error